wgpu/src/image/raster.rs

```diff
@@ -153,10 +61,45 @@ impl Cache {
         self.get(handle).unwrap()
     }
 
-    pub fn trim(&mut self) {
+    pub fn upload(
+        &mut self,
+        handle: &image::Handle,
+        device: &wgpu::Device,
+        encoder: &mut wgpu::CommandEncoder,
+        atlas: &mut Atlas,
+    ) -> Option<&atlas::Entry> {
+        let memory = self.load(handle);
+
+        if let Memory::Host(image) = memory {
+            let (width, height) = image.dimensions();
+
+            let entry = atlas.upload(width, height, &image, device, encoder)?;
+
+            *memory = Memory::Device(entry);
+        }
+
+        if let Memory::Device(allocation) = memory {
+            Some(allocation)
+        } else {
+            None
+        }
+    }
+
+    pub fn trim(&mut self, atlas: &mut Atlas) {
         let hits = &self.hits;
 
-        self.map.retain(|k, _| hits.contains(k));
+        self.map.retain(|k, memory| {
+            let retain = hits.contains(k);
+
+            if !retain {
+                if let Memory::Device(entry) = memory {
+                    atlas.remove(entry);
+                }
+            }
```

Comment:
I reverted the smart deallocation on drop, as it added a bunch of `Rc` and `RefCell` complexity. I think we should keep deallocation centralized and easy to reason about.

Reply:
Wasn't sure about that on either. Automatic deallocation just seemed kind of "rusty".